pmlc/dialect/pxa/ir/ops.cc

```diff
@@ -89,6 +113,50 @@ AffineParallelOp::operand_range AffineParallelOp::getUpperBoundsOperands() {
   return {operand_begin() + lowerBoundsMap().getNumInputs(), operand_end()};
 }
 
+size_t AffineParallelOp::getNumDims() {
+  return steps().cast<ArrayAttr>().size();
+}
+
+AffineValueMap AffineParallelOp::getLowerBoundsValueMap() {
+  llvm::SmallVector<Value, 8> ops;
+  for (auto op : getLowerBoundsOperands()) {
+    ops.push_back(op);
+  }
+  return AffineValueMap(lowerBoundsMap(), ops);
+}
+
+AffineValueMap AffineParallelOp::getUpperBoundsValueMap() {
+  llvm::SmallVector<Value, 8> ops;
+  for (auto op : getUpperBoundsOperands()) {
+    ops.push_back(op);
+  }
+  return AffineValueMap(upperBoundsMap(), ops);
+}
+
+AffineValueMap AffineParallelOp::getRangesValueMap() {
+  AffineValueMap out;
+  AffineValueMap::difference(getUpperBoundsValueMap(), getLowerBoundsValueMap(),
+                             &out);
+  return out;
+}
+
+bool AffineParallelOp::getConstantRanges(llvm::SmallVectorImpl<int64_t> &out) {
```

Comment:
Return an `Optional<SmallVector<int64_t, N>>` here instead of `bool` + out param.

Reply:
Done.